rancher/k3s
# Galal hussein etcd backup restore

Proposed changes
======

Add etcd snapshot and restoration feature, this include adding 5 new flags:

- "etcd-snapshot-interval"
- "etcd-snapshot-dir"
- "etcd-snapshot-restore-path"
- "etcd-disable-snapshots"
- "etcd-snapshot-retention"

Types of changes
======

- New feature (non-breaking change which adds functionality)

Verification
======

snapshot should be enabled by default and the snapshot dir defaults to <data-dir>/server/db/snapshots, you should verify by checking the directory for snapshots saved there.

restoration should happen the same way cluster-reset is working, when specifying the flag it will move the old data dir to <data-dir>/server/db/etcd-old and then attempt to restore the snapshot by creating a new data dir, and then start etcd with forcing a new cluster so that it start a 1 member cluster. Upon completion of the restore, k3s will exit.

to verify the restoration you should be able to see cluster starting with only one etcd member and the data from the specified snapshot is being restored correctly.

The pr also will add etcd snapshot retention and also a flag to disable snapshots altogether.


Testing
======

###1 Testing disabling snapshots

- start k3s with --cluster-init and --etcd-disable-snapshots

you should see no snapshot has been created in <data-dir>/server/db/snapshots

###2 Testing snapshot interval and retention

- start k3s with --etcd-snapshot-interval set to 5s and --etcd-snapshot-retention set to 10

you should see snapshots created every 5 seconds in <data-dir>/server/db/snapshots and no more than 10 snapshots in the directory

Make sure that if the `--cluster-reset-restore-path` flag is given that the `--cluster-reset` flag is present:

```sh
./k3s server --datastore-endpoint=etcd --cluster-reset-restore-path=/home/bdowns/etcd_snapshot-etcd-snapshot-1598631420
FATA[2020-08-28T19:06:27.534957919Z] Invalid flag use. --cluster-reset required with --cluster-reset-restore-path
```

Verify that there's a failure when an invalid path is provided:

```sh
FATA[2020-08-28T19:07:46.056232602Z] starting kubernetes: preparing server: start cluster and https: etcd: snapshot path does not exist: /home/bdowns/etcd_snapshot-etcd-snapshot-1598631420
```

###3 testing snapshot restore

To test snapshot restore, just use --etcd-snapshot-restore-path and point to any snapshot file that you have on the system

The cluster should restore and etcd will only have one member, also you should see <data-dir>/server/db/etcd-old directory has been created.

Linked Issues
======

https://github.com/rancher/rke2/issues/45
